src/Provider/AbstractProvider.php

```diff
@@ -98,7 +108,15 @@ public function __construct($options = [], HttpAdapterInterface $httpClient = nu
             }
         }
 
-        $this->setHttpClient($httpClient ?: new CurlHttpAdapter());
+        if (empty($collaborators['httpClient'])) {
+            $collaborators['httpClient'] = new CurlHttpAdapter();
+        }
+        $this->setHttpClient($collaborators['httpClient']);
+
+        if (empty($collaborators['randomFactory'])) {
+            $collaborators['randomFactory'] = new RandomFactory();
+        }
+        $this->setRandomFactory($collaborators['randomFactory']);
```

Comment:
The factory is far more flexible, and you can register any number of mixers and sources with it. By having the factory inside the class, providers could make use of it for their own needs, instead of being limited to a single generator.